akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -269,6 +270,57 @@ object GraphStages {
   def terminationWatcher[T]: GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]] =
     TerminationWatcher.asInstanceOf[GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]]]
 
+  private class FlowMonitorImpl[T] extends FlowMonitor[T] {
+    private val stateRef = new AtomicReference[StreamState[T]](Initialized)
+    def setState(state: StreamState[T]): Unit = stateRef.set(state)
+
+    override def state = stateRef.get()
+  }
```

Comment:
What happens if T == StreamState? Are we OK with that?
I mean, it would be a shame to miss out on this optimization for such an edge case, but it still feels a bit problematic.